src/trace-plugin-loader.ts

```diff
@@ -15,26 +15,109 @@
  */
 'use strict';
 
+// TODO(kjin)
+// Module augmentation is implemented in this PR:
+// https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19612
+// Do this correctly when it's landed and released.
+import _Module = require('module');
+const Module: {
+  _resolveFilename(request: string, parent?: NodeModule): string;
+  _load(request: string, parent?: NodeModule, isMain?: boolean): any;
+} = _Module as any;
+
+import { Logger } from '@google-cloud/common';
+import * as path from 'path';
+import * as semver from 'semver';
+import * as shimmer from 'shimmer';
 import * as util from './util';
-import { TraceAgent } from './trace-api';
+import { TraceAgent, TraceAgentConfig } from './trace-api';
+
+/**
+ * An interface representing config options read by the plugin loader, which includes
+ * TraceAgent configuration as well.
+ */
+export interface PluginLoaderConfig extends TraceAgentConfig {
+  plugins: {
+    [pluginName: string]: string;
+  };
+}
+
+export interface Patch<T> {
+  file?: string;
+  versions?: string;
+  patch: (module: T, agent: TraceAgent) => void;
+  unpatch?: (module: T) => void;
+}
+
+export interface Intercept<T> {
+  file?: string;
+  versions?: string;
+  intercept: (module: T, agent: TraceAgent) => T;
+}
+
+export type Instrumentation<T> = Patch<T> | Intercept<T>;
+
+export type Plugin = Array<Instrumentation<any>>;
+
+interface InternalPatch<T> extends Patch<T> {
+  file: string;
+  module?: T;
+  patch: (module: T, agent: TraceAgent) => void;
```

Comment:
can patch & unpatch definitions be omitted? The should get inherited from Patch.